### ISSUE TITLE
Support module.yml in zephyr repo

### ISSUE THREAD

**Issue by gregshue** [zephyrproject-rtos/zephyr#42700]
**Is your enhancement proposal related to a problem? Please describe.**
The `zephyr` repository is the only one in the standard manifest that is not managed as an actual module.  This prevents using the same **algorithmic** solution for referencing files in Zephyr from another module as used to reference other modules.  

This "special case" handling of finding the zephyr-resident code tree is contrary to best practices in software engineering, makes all related logic more complicated, and increases the cost of understanding the Zephyr build system when debugging issues.

As a down-stream integrator, I would also like the freedom to tweak my fork of Zephyr just like every other module.

**Describe the solution you'd like**
I would like to see the zephyr build system fully support having a "module.yml" file in the zephyr repo and treat it exactly like every other module.

**Describe alternatives you've considered**
I have not considered any other alternatives, since this is a request for simplification.

**Additional context**
None.

**Comment by hongshui3000** [comment]
BTW,The documentation section would ideally also be a module, the documentation is large, and not everyone needs to rebuild the documentation or unstructured documentation source material.

**Comment by gregshue** [comment]
Actually, the Zephyr-structured documentation section(s) should be discovered through module.yml as well, leaving  style, logo, etc to be picked up from wherever the doc build was started. 😉

Sent from my iPhone

On Feb 10, 2022, at 10:28 PM, yxh ***@***.***> wrote:

﻿

BTW,The documentation section would ideally also be a module, the documentation is large, and not everyone needs to rebuild the documentation or unstructured documentation source material.

**Comment by tejlmand** [comment]
> The zephyr repository is the only one in the standard manifest that is not managed as an actual module.

Because it is not a module, it's a lot more.

> This prevents using the same algorithmic solution for referencing files in Zephyr from another module as used to reference other modules.

Exactly because it is not a module, so being able to refer to Zephyr using `ZEPHYR_ZEPHYR_MODULE_DIR` would give the wrong impression.

This is no different than the fact that if you include a repository in your manifest, which contains code but is not a Zephyr module, then you cannot refer to it as `ZEPHYR_<module>_MODULE_DIR`.

For example, like in this case: https://github.com/nrfconnect/sdk-nrf/blob/cfedfdfa08567b2252b511a4d1db15fbeba8152d/west.yml#L156-L160
```
    - name: mbedtls-nrf
      path: mbedtls
      repo-path: mbedtls
      revision: mbedtls-2.26.0
      remote: armmbed
```
which include [ARM mbed TLS](https://github.com/ARMmbed/mbedtls/tree/v2.26.0) but it's not a Zephyr module.
You have to refer to code in such a repository in another way, not as a module.

Now, making Zephyr itself a module bears the risk of complicating the build flow and confuse users.
Clearly not all code can be treated as module code, cause all the logic around creating the core build system parts, setting up toolchains, handling of Zephyr modules, etc. must be done before any Zephyr module can be loaded.

Whereas other parts of the build system in principle can be loaded as module code, for example `subsys` and `drivers`.
I'm concerned such split can also cause confusion, cause suddenly some parts of the Zephyr tree is loaded as core build parts and other parts are loaded as module code.

This will make it hard for users to keep track of which code parts are loaded as core build part and which are loaded as module.

I do see some pros in this proposal and is not opposed to it, but I also have some doubts and concerns.
Anyone with an opinion on this matter, feel free to drop a comment.

**Comment by gregshue** [comment]
> it's a lot more.

Agreed - partially.  It is also:
- a manifest repository (also true of `example-application` module)
- contains the build/document generation/test runner systems (which eventually need to be extensible)

Once we trim the Zephyr directory structure down to what is not currently accessible via `module.yml` and is not repository-specific we end up with:

- `cmake/`  (build/tools)
- `doc/` (which will get revisited when we start enumerating user-level requirements)
- `Kconfig.zephyr`  (build/tools - which would have some content moved into existing `Kconfig`)
- `MAINTAINERS.yml`  (codebase metadata)
- `scripts/`  (build/tools)
- `share/`  (build/tools)
- `zephyr-env.*`  (build/tools)

I disagree that this is "a lot" more.

What am I not noticing?

> making Zephyr itself a module bears the risk of complicating the build flow 

How does this complicate the build flow?

> This will make it hard for users to keep track of which code parts are loaded as core build part and which are loaded as module.

If I ever needed to care (which I haven't needed to) then I look at how it is referenced.

-  `ZEPHYR_BASE` == install/build/doc generation tools
- `$workspace/zephyr/zephyr/module.yml` == module
- `$workspace/zephyr/**/*.[sh, py, bat, etc]` == tools

When will I need to care in a way not covered by the manner of reference?

**Comment by carlescufi** [comment]
> This prevents using the same **algorithmic** solution for referencing files in Zephyr from another module as used to reference other modules.

@gregshue in order for the benefits of this proposal to be clearer to all, could you expand on what this means in practice please?

**Comment by tejlmand** [comment]
> > making Zephyr itself a module bears the risk of complicating the build flow
>
> How does this complicate the build flow?

Today, both Zephyr modules and `west` are optional.

With this idea you will have to always get Zephyr in as a module, and if using `west` this would then happen automatically but if not using `west` are users then suddenly expected to include Zephyr as a module ?
And if not, and Zephyr always includes itself as a module also, then how are users supposed to opt-out of this module.
The whole idea behind Zephyr modules, is to make it possible to add / remove modules to/from Zephyr build system depending on your needs.

It is not intended as a way to load Zephyr itself.

Now, starting to treat parts of the Zephyr repo itself this ways opens up for lots of questions as to what a module is.
How it is handled / configured (notice how modules Kconfigs are sourced into a dedicated tree in Kconfig).
What parts of Zephyr are considered to be so optional that users should have the ability to opt-out ? network, bluetooth, logging, kernel, ...

**Comment by gregshue** [comment]
> if not using west are users then suddenly expected to include Zephyr as a module ?
No. Even without `west`, the solution can automatically add Zephyr to the module list if there is not already a module named 'zephyr'.

> how are users supposed to opt-out of this module.
It is impractical for users to opt-out of the module providing standard/infrastructure interface headers, even if they opted out of every _implementation_ (e.g., libs/drivers/subsystems/tests/samples/etc) in the Zephyr repository.  As indicated above, a user _could_ provide their own module named 'zephyr' if another solution is not provided.

At this point we do not have a standard way to "opt-out" of a module other than removing it from the module list.  If every module defined a boolean `CONFIG_<MODULE>` and made all its CMakeLists.txt content dependent on it, then we would have a standard solution.

> The whole idea behind Zephyr modules, is to make it possible to add / remove modules to/from Zephyr build system depending on your needs.
As a user, I also need the ability to add/remove/replace implementations of individual subsystems/components/drivers rather than modules.  This is true even for the implementations in Zephyr.  Extending the Zephyr repo to be a module actually aligns needs and drives toward a standard solution.  

> How [the Zephyr module] is handled / configured?
The build system configuration stays as is.  The source trees that are found in other modules are handled exactly the same way here.

> What parts of Zephyr are considered to be so optional that users should have the ability to opt-out ? network, bluetooth, logging, kernel, ...
It does not matter whether they opt out or replace with an alternate implementation (to fix bugs or provide enhancements or ...). That is not for us to decide. It is up to us to provide the capability in a _reusable_ release.  (BTW, yes, yes, yes, yes, ... essentially every implementation, but not interface definitions or configuration settings.)

> Zephyr modules are not guaranteed to be processed in specific order ... 
That issue is independent of this one.  Other modules will be found in similar circumstances as Zephyr does.

> if Zephyr is being included by another manifest file, then it might be sourced later
`when`, not `if`.  That issue is independent of this one, and already exists.

> then it might be sourced later, will that work?
That is an existing issue.  This proposal just highlights the need.

> must all modules suddenly depend on Zephyr?
They already do.  Consider the dependencies for at least the `module.yml` definition, and likely for the Zephyr wrappers to CMake commands,  Kconfig support/definitions/conversions to preprocessor symbols.

**Comment by gregshue** [comment]
Though in the original issue I only mentioned one reason for extending the Zephyr repository to be a module, there are several.  Consider:
- Following the (best practice) Software Engineering Design Principle of Separation of Concerns (http://swebokwiki.org/Chapter_2:_Software_Design#Software_Design_Principles). Clearly separating the portions elsewhere managed in modules from the build system is a broadly accepted strategy in managing complexity. This is a demonstration of commitment to stated values and modeling desired behavior.
- Highlighting the broad support of the `module.yml` interface for extensibility. This is already a requirement, and implementation and automated verification have been lacking.
- Raising increased awareness that issues and enhancements relevant to the module `zephyr` are almost(?) always relevent to other modules, and vice-versa.  This reduces the risk of oversights in related build system architecture and functionality.
- Providing a reference (and test case) for when modules need to build subdirectories located above the provided `zephyr/CMakeLists.txt`.

**Comment by tejlmand** [comment]
> As a user, I also need the ability to add/remove/replace implementations of individual subsystems/components/drivers rather than modules. 

and this is a part to which I agree, we should strive to have Zephyr as modular as possible when it's possible, and not making working with Zephyr unnecessary complex.
However, being able to easily integrate with Zephyr and adjust it to custom needs is **not** the same as making Zephyr itself a module.
For example, Zephyr already offers various possibilities for customization, for example:
https://github.com/zephyrproject-rtos/zephyr/blob/22c0843485b5d03a8bf1de8499fe1d20d2ded3d3/drivers/entropy/Kconfig.nrf5#L7-L12
https://github.com/zephyrproject-rtos/zephyr/blob/22c0843485b5d03a8bf1de8499fe1d20d2ded3d3/cmake/extensions.cmake#L455

And I believe it would be great if such patterns could be better formalized and used in a consistent way, so that it is easier and more predictable when / where / how Zephyr can be customized.

> > What parts of Zephyr are considered to be so optional that users should have the ability to opt-out ? network, bluetooth, logging, kernel, ...
>
> It does not matter whether they opt out or replace with an alternate implementation

It was intended as a rhetorical question, to show that those parts are core to Zephyr.
Of course a user can decide not to enable them, but they are core in the sense they should always be available to any sample which might want to enable them.
With my rhetorical question, I was simply trying to show why Zephyr itself is not a module. 



> > Zephyr modules are not guaranteed to be processed in specific order ...
>
> That issue is independent of this one.
> > if Zephyr is being included by another manifest file, then it might be sourced later
>
> That issue is independent of this one, and already exists.
> > then it might be sourced later, will that work?
>
> That is an existing issue. This proposal just highlights the need.
> > must all modules suddenly depend on Zephyr?
>
> They already do.

No, those issues doesn't exist today cause Zephyr is not a module but the foundation on which all the other modules builds upon.
And this exactly highlights why Zephyr itself is not to be considered a module.


So to sum up, my current state on this:

Zephyr could improve in following areas:
- Improved support for customization, today some `_ALT` settings and cmake commands exists to allow tweaking, but those are introduced on a need-to-have basis and grown organically.
   It might be time to formalize a practice in this area
- Improved modularization for reuse and support easier integration to the Zephyr build system, work is already happening in the area, see #41301
- Doc building and possibility to customize this seems like a valid request, however I would consider such improvement low priority compared to many other enhancement tasks pending for the build system.

but Zephyr is **not** a module, and should not be treated as such.
So if you change the purpose of this issue to better integration. reuse, modularization, etc. of Zephyr, then I would be happy.

**Comment by gregshue** [comment]
> we should strive to have Zephyr as modular as possible when it's possible, and not making working with Zephyr unnecessary complex.

Yes. I completely agree with the goal.

> being able to easily integrate with Zephyr and adjust it to custom needs is not the same as making Zephyr itself a module.

Yes. I completely agree.

> It was intended as a rhetorical question, to show that those parts are core to Zephyr.

It is actually a very pertinent question, which is why I responded to it. :-)

> they are core in the sense they should always be available to any sample which might want to enable them.

Be very careful here.  If we use this criteria for "core", then all HALs (indeed every module) are also core.  That is probably not what you intended. There are enough dependencies each direction between things in the Zephyr repo and the existing modules that (almost) everything becomes "core".

--

So I carefully re-read [Module Repositories](https://docs.zephyrproject.org/latest/guides/modules.html?highlight=module#module-repositories).  It seems the bullets at the beginning of this section are the normative definition of a module. 

First note:

> - If the module has an external (upstream) project repository ...

The "If" clearly indicates modules do not have to be external projects.  Indeed, Zephyr's own [example-application](github.com/zephyrproject-rtos/example-application) is a module that is not an external project. 

Let's evaluate your concerns against this definition:

> Zephyr modules are not guaranteed to be processed in specific order 

> Ref: https://docs.zephyrproject.org/latest/guides/modules.html#zephyr-module-dependencies

> must all modules suddenly depend on Zephyr ?

> Zephyr is ... the foundation on which all the other modules builds upon

It seems the concern is having to retrofit existing modules to describe the implicit dependencies on the Zephyr repository that already exist. This is an important concern, and would need to be addressed. 

 _As a concession to backwards compatibility_, the `zephyr_module.py` could be changed from stripping `zephyr` from the module list to ensuring `zephyr` occurs first in the module list (as already proposed).

> if Zephyr is being included by another manifest file, then it might be sourced later

Regarding the `zephyr` repository, it seems ensuring `zephyr` occurs first in the module list would address that concern.

> And this exactly highlights why Zephyr itself is not to be considered a module

With the concession for backwards compatibility previously identified, it seems:
- the stated concerns are addressed
- we have taken a step towards making Zephyr "as modular as possible when it's possible" 
- we have reduced complexity by making implicit layers explicit and follow a pattern consistent with analogous content in other modules, thus making it "more predictable when / where / how Zephyr can be customized"
- we are making a change consistent with applying best practices (a stated value of the Zephyr Project)
- we have simplified how to describe the `zephyr` repository by changing it from an "is-a" (polymorphic - where everything in it is accessed via ZEPHYR_BASE) to a "has-a" (composition - where some content is accessed via ZEPHYR_ZEPHYR_MODULE_DIR)-

Are there any new concerns?

**Comment by tejlmand** [comment]
It has been discussed several times that Zephyr is not, and shall not be a module.
Zephyr is Zephyr, the core of the system.

See also: https://github.com/zephyrproject-rtos/zephyr/pull/42927
which has been rejected.

**Comment by nordicjm** [comment]
> As a down-stream integrator, I would also like the freedom to tweak my fork of Zephyr just like every other module.

Then I suggest you make these changes downstream. Fork zephyr and make your changes to your own fork, you have the freedom to make whatever changes you want in such a fork, call it something like gregphyr and make it a module. As for upstream, this is a NACK from me, the whole idea is completely at odds with zephyr.
